core/system/inc/context.h

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2013-2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef __CONTEXT_H__
+#define __CONTEXT_H__
+
+#include "api/inc/context_exports.h"
+#include "api/inc/uvisor_exports.h"
+#include <stdint.h>
+
+/** Exception stack frame size
+ *
+ * @warning Currently only the FPU-disabled case is supported. Exceptions that
+ * generates when the FPU is enabled will escalate to a hard fault. */
+#define CONTEXT_SWITCH_EXC_SF_SIZE 8
+
+/** Maximum number of arguments that can be passed to a function-bound context
+ * switch. */
+#define CONTEXT_SWITCH_FUNCTION_MAX_NARGS 4
+
+/** Types of context switches
+ *
+ * There are two main types of context switches:
+ *  1. Tied to a function. Returning from the context switch will trigger a
+ *     callback to a destination function. Arguments can be passed in the
+ *     process.
+ *  2. Unbound. Only the state and MPU configurations are switched. The host OS
+ *     must take care of deferring execution to a different callback. The stack
+ *     frame and stack pointer of the destination box are not touched. */
+typedef enum {
+    CONTEXT_SWITCH_FUNCTION_GATEWAY, /**< Bound to function. Allows 4 32 bit arguments, 1 32 bit return value. */
+    CONTEXT_SWITCH_FUNCTION_ISR,     /**< Bound to an ISR. Does not allow arguments, nor a return value. */
+    CONTEXT_SWITCH_FUNCTION_DEBUG,   /**< Bound to debug handler. Uses the format specified by the debug box driver. */
+    CONTEXT_SWITCH_UNBOUND_THREAD,   /**< Not bound to a handler. Used for thread switching. No switch-out. */
+} TContextSwitchType;
+
+/** Snapshot of a previous context state
+ *
+ * This struct contains the information needed to restore a previously stalled
+ * context. It is needed when returning from a nested context switch.
+ *
+ * @warning We assume that the context type (a value from the
+ * ::TContextSwitchType enum) fits into 8 bits. */
+typedef struct {
+    uint8_t __pad[2]; /**< Padding to get 32 bit alignment */
```

Comment:
Good point! I changed it. I though it didn't make much difference, but actually a single `ldrb` instruction to the struct location fetches the data directly (no need of a second load). That's also why I have now put `src_id` first.